rgb() colors that use none should use hex syntax for serialization in canvas

weinig opened the pull request:
#### b7e7afad1304ab572e0b27562c4b0c9747930ad8
<pre>
rgb() colors that use none should use hex syntax for serialization in canvas
<a href="https://bugs.webkit.org/show_bug.cgi?id=278392">https://bugs.webkit.org/show_bug.cgi?id=278392</a>

Reviewed by Darin Adler.

When using HTML serialization (such as in canvas code), `rgb(none none none)`
should serialize as `#000000`, not `rgb(0, 0, 0)` as it currently does.

This is more consistent and matches other browsers.

* LayoutTests/fast/canvas/canvas-color-serialization.html:
* LayoutTests/fast/canvas/canvas-color-serialization-expected.txt:
    - Add serialization tests. Additional will be added in WPT
      via <a href="https://github.com/web-platform-tests/wpt/pull/47148.">https://github.com/web-platform-tests/wpt/pull/47148.</a>

* Source/WebCore/platform/graphics/ColorSerialization.cpp:
(WebCore::serializationForHTML):
    - Make sure to call through to the HTML serializer for legacy
      rgb() colors that have none (which is indicated by the use
      of SRGBA&lt;float&gt; without the &quot;useColorFunctionSerialization&quot;
      bit set).

Canonical link: <a href="https://commits.webkit.org/283120@main">https://commits.webkit.org/283120@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2f5c40452665d6b66d27837c74bca2426808109b

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65265 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44634 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/17879 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69289 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/15871 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/67383 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52416 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16153 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52423 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/10986 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68331 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56480 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33046 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/37916 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/13851 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14748 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/59793 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14191 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/70994 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9217 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13674 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59749 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9249 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56541 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60023 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/ignored-objects (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14389 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7610 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/1311 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40444 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41521 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/42702 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41265 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/platform/graphics/ColorSerialization.cpp:
```diff
@@ -553,7 +553,10 @@ String serializationForCSS(const SRGBA<float>& color, bool useColorFunctionSeria
+    if (useColorFunctionSerialization)
+        return serializationUsingColorFunction(color);
+
+    return serializationForHTML(convertColor<SRGBA<uint8_t>>(color), false);
```
Is there any issue with loss of precision here if values are not integers?

weinig replied:
No, this is what the spec requires. When precision is actually wanted, the new syntaxes (e.g. color(srgb ...)) maintain full precision. This is just weird legacy edge cases.

weinig committed:
rgb() colors that use none should use hex syntax for serialization in canvas
https://bugs.webkit.org/show_bug.cgi?id=278392

Reviewed by Darin Adler.

When using HTML serialization (such as in canvas code), `rgb(none none none)`
should serialize as `#000000`, not `rgb(0, 0, 0)` as it currently does.

This is more consistent and matches other browsers.

* LayoutTests/fast/canvas/canvas-color-serialization.html:
* LayoutTests/fast/canvas/canvas-color-serialization-expected.txt:
    - Add serialization tests. Additional will be added in WPT
      via https://github.com/web-platform-tests/wpt/pull/47148.

* Source/WebCore/platform/graphics/ColorSerialization.cpp:
(WebCore::serializationForHTML):
    - Make sure to call through to the HTML serializer for legacy
      rgb() colors that have none (which is indicated by the use
      of SRGBA<float> without the "useColorFunctionSerialization"
      bit set).

Canonical link: https://commits.webkit.org/283120@main
